Implement hash and HMAC stream one shots

vcsjones opened the pull request:
This implements hashing and HMAC statics for streams. Additionally,
"LiteHmac" and "LiteHash" were introduced. The existing HMAC and hash
provider functionality do some bookkeeping we don't need for resetting.
Since we do not need to use these hash handles after the digest has
been finalized, resetting is unnecessary work. For HMAC, that also means
keeping a copy of the key around for some implementations which we don't
need to do.

The LiteHash and LiteHmac types are implemented as structs with a common
interface. To avoid boxing, generics are used and constrained to the interface
where possible.

The Browser implementation just defers to the existing HashDispenser rather
than do anything novel.

Closes #62489.

vcsjones committed:
Implement hash and HMAC stream one shots.

This implements hashing and HMAC statics for streams. Additionally,
"LiteHmac" and "LiteHash" were introduced. The existing HMAC and hash
provider functionality do some bookkeeping we don't need for resetting.
Since we do not need to use these hash handles after the digest has
been finalized, resetting is unnecessary work. For HMAC, that also means
keeping a copy of the key around for some implementations which we don't
need to do.

The LiteHash and LiteHmac types are implemented as structs with a common
interface. To avoid boxing, generics are used and constrained to the interface
where possible.

The Browser implementation just defers to the existing HashDispenser rather
than do anything novel.

vcsjones commented:
Draft to get feedback from CI, and this is going to conflict with #63685. That should be merged first and I will resolve conflicts.

vcsjones commented:
This does not implement any shortcuts for some streams. We could possibly do this for `MemoryStream` (`TryGetBuffer`) and `UnmanagedMemoryStream` (`PositionPointer`). That can be done in a follow-up PR rather than make this PR bigger.

stephentoub commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACMD5.cs:
```diff
@@ -155,6 +158,181 @@ public static bool TryHashData(ReadOnlySpan<byte> key, ReadOnlySpan<byte> source
+
+            if (!source.CanRead)
+                return ValueTask.FromException<byte[]>(
+                    new ArgumentException(SR.Argument_StreamNotReadable, nameof(source)));
```
We typically don't wrap user errors (e.g. argument exceptions) in tasks.

vcsjones replied:
Ha, that makes sense. Fixed.

vcsjones committed:
Code review feedback.

- Throw exceptions synchronously.
- Dispose of hash even if CryptoPool rent/return fails.

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/LiteHashProvider.cs:
```diff
@@ -0,0 +1,191 @@
+        /// This takes ownership of the hash parameter and disposes of it when done.
+        private static int ProcessStream<T>(T hash, Stream source, Span<byte> destination) where T : ILiteHash
+        {
+            byte[] rented = CryptoPool.Rent(4096);
```
I don't think it would matter in practice, the safe handles would get finalized. But I don't see a reason why we can't address it anyway, so, fixed.

vcsjones committed:
Merge remote-tracking branch 'ms/main' into hash-stream-oneshot-attempt-2

vcsjones committed:
Fix compilation from merged changes

bollhals commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/LiteHashProvider.cs:
```diff
@@ -0,0 +1,195 @@
+        }
+
+        /// This takes ownership of the hash parameter and disposes of it when done.
+        private static int ProcessStream<T>(T hash, Stream source, Span<byte> destination) where T : ILiteHash
```
I've done something similar (in readonly struct + generic constrained to interface) in a library I contribute to and while it works, any method called on that interface will force a defensive copy, since the compiler is not able to recognize that the struct method itself is readonly. I think it only sees the interface and thus decides on a defensive copy. I'm also not sure anymore whether without the in modifier as you have here, the compiler is able to figure it out properly or not.

Based on the size of the structs used, I assume it's not as relevant here, just wanted to mention it.

vcsjones committed:
Merge remote-tracking branch 'ms/main' into hash-stream-oneshot-attempt-2

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACSHA256.cs:
```diff
@@ -163,6 +165,176 @@ public static bool TryHashData(ReadOnlySpan<byte> key, ReadOnlySpan<byte> source
+            CancellationToken cancellationToken = default)
+        {
+            ArgumentNullException.ThrowIfNull(source);
+
+
```
Nit: Too many blank lines

```suggestion

```

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/LiteHash.Apple.cs:
```diff
@@ -0,0 +1,163 @@
+        }
+    }
+
+    internal readonly struct LiteHash : ILiteHash
```
LiteHash looks a lot like HashProvider.  Why do we need both?

vcsjones replied:
* LiteHmac doesn't keep a key copy for resets
* LiteHash / LiteHmac don't reset once used.
* LiteHash / LiteHmac don't keep track of whether operations are in-flight with `_running`
* It's a struct

_Could we put a `Reset` on these (taking the key for the HMAC variant) and have `HashProvider` wrap those?_

Maybe. That's kind of what I did for the symmetric one-shots (replacing IV for HMAC keys). I avoided that here because I felt that the structs were simple enough on their own, really just a few methods on a `SafeHandle`. De-duplicating it would just mean another level of indirection on the HashProviders to the Lite ones.

bartonjs replied:
Hm.  Feels like the HashProvider could just have a LiteHash as a field, and then manage the state on top of that.

If it doesn't nest easily, then it's not a whole lot of duplication... but it's still nice to only need to make changes N times (Windows, Unix, Android, Apple OSes, Browser) instead of N * M (when M > 1)... so I have to ask :smile:

vcsjones replied:
This "works" for Unix / Android and Apple. It works less-so to me for Windows. The Windows one was already somewhat specialized, and it does upfront work to determine if the platform supports usable CNG hashes.

What didn't work either was a single `HashProvider` for multiple platforms, even if they wrapped their own `LiteHash`, since everyone wants their own state.

If the last two commits are just seem like noise without pushing far enough in to the "make it less duplicated", I can revert or drop them.

The alternative, if we really do not like the duplication is to make the stream one shots allocate a bunch more and abandon the lite hashes altogether. Though I believe the one-shots should remain as light as reasonably possible, even if they do end up allocating a safe-handle or two.

bartonjs replied:
While "no duplication" would be ideal, "less duplication" is at least virtuous.  Thanks for pursuing it.

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/LiteHash.Windows.cs:
```diff
@@ -0,0 +1,173 @@
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+using NTSTATUS = Interop.BCrypt.NTSTATUS;
+using BCryptOpenAlgorithmProviderFlags = Interop.BCrypt.BCryptOpenAlgorithmProviderFlags;
+using BCryptCreateHashFlags = Interop.BCrypt.BCryptCreateHashFlags;
```
Sorted looks better to my eye.

```suggestion
using BCryptCreateHashFlags = Interop.BCrypt.BCryptCreateHashFlags;
using BCryptOpenAlgorithmProviderFlags = Interop.BCrypt.BCryptOpenAlgorithmProviderFlags;
using NTSTATUS = Interop.BCrypt.NTSTATUS;
```

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/MD5.cs:
```diff
@@ -112,6 +115,133 @@ public static bool TryHashData(ReadOnlySpan<byte> source, Span<byte> destination
+        public static ValueTask<byte[]> HashDataAsync(Stream source, CancellationToken cancellationToken = default)
+        {
+            if (source is null)
+                return ValueTask.FromException<byte[]>(new ArgumentNullException(nameof(source)));
```
The HMACs all threw these as preconditions instead of results.  Preconditions seems correct.

```suggestion
                throw new ArgumentNullException(nameof(source));
```

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/MD5.cs:
```diff
@@ -112,6 +115,133 @@ public static bool TryHashData(ReadOnlySpan<byte> source, Span<byte> destination
+            if (source is null)
+                return ValueTask.FromException<byte[]>(new ArgumentNullException(nameof(source)));
+
+            if (!source.CanRead)
+                return ValueTask.FromException<byte[]>(
+                    new ArgumentException(SR.Argument_StreamNotReadable, nameof(source)));
```
throw

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/MD5.cs:
```diff
@@ -112,6 +115,133 @@ public static bool TryHashData(ReadOnlySpan<byte> source, Span<byte> destination
+            Memory<byte> destination,
+            CancellationToken cancellationToken = default)
+        {
+            if (source is null)
+                return ValueTask.FromException<int>(new ArgumentNullException(nameof(source)));
+
+            if (destination.Length < HashSizeInBytes)
+                return ValueTask.FromException<int>(
+                    new ArgumentException(SR.Argument_DestinationTooShort, nameof(destination)));
+
+            if (!source.CanRead)
+                return ValueTask.FromException<int>(
+                    new ArgumentException(SR.Argument_StreamNotReadable, nameof(source)));
```
throw :smile:

vcsjones replied:
🤔 thought I fixed all of these from Steve's feedback. Maybe I didn't push something.

bartonjs replied:
Looks like only MD5 got missed.  Probably just missed it as the only hash to not start with S.

vcsjones replied:
Ah I missed MD5 because I'm a dummy.

bartonjs commented on src/libraries/System.Security.Cryptography/tests/HashAlgorithmTestDriver.cs:
```diff
@@ -320,6 +374,85 @@ protected void VerifyRepeating(string input, int repeatCount, string output)
+        [Fact]
+        public async Task HashDataAsync_Null_Stream_Throws()
+        {
+            await AssertExtensions.ThrowsAsync<ArgumentNullException>(
+                "source",
+                async () => await HashDataAsync((Stream)null, cancellationToken: default));
```
The argument validation ones can/should be synchronous.

```suggestion
            Assert.Throws<ArgumentNullException>(
                "source",
                HashDataAsync((Stream)null, cancellationToken: default));
```

bartonjs commented on src/libraries/System.Security.Cryptography/tests/Sha384Tests.cs:
```diff
@@ -42,7 +118,7 @@ public void Sha384_NistShaAll_1()
 
         [Fact]
-        public void Sha256_Fips180_MultiBlock()
+        public void SHA384_Fips180_MultiBlock()
```
Why change the casing here?  All of the rest of the methods are little-H/little-A

vcsjones replied:
Ah, well, it was called out as SHA**256** instead of SHA**384** and I guess my fingers just did the wrong thing while fixing the algorithm size. Will fix.

bartonjs replied:
Ah.  Clearly my internal diffing algorithm stopped at the first change. :)

vcsjones committed:
Fix MD5 to throw exceptions synchronously

vcsjones committed:
Fix usings

vcsjones committed:
Fix extra white space

vcsjones committed:
Code review feedback for tests.

vcsjones committed:
Update HashDispensers to use Lite hashes.

* OpenSsl / Android can wrap.
* Apple can wrap.
* Windows should not wrap. The HashProviderCng is somewhat specialized in its ability to reset. It did up-front check to determine if the platform supported reusable hash providers, and further had a single implementation for HMAC and Digests. The current Lite hash design requires that they remain separate types.

vcsjones committed:
Don't double-initialize macOS HMAC
